libc/mman.c

```diff
@@ -0,0 +1,355 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/utils.h>
+#include <stdlib.h>
+
+#include "mman.h"
+#include "openenclave/bits/defs.h"
+#include "syscall.h"
+
+static oe_mapping_t* _mappings;
+static oe_spinlock_t _lock;
+
+static void _clear_mappings(void)
+{
+    oe_mapping_t* m = _mappings;
+    _mappings = 0;
+    while (m)
+    {
+        oe_mapping_t* next = m->next;
+        free((void*)m->start);
+        free(m->status_vector);
+        free(m);
+        m = next;
+    }
+}
+
+static void _call_atexit(void)
+{
+    atexit(_clear_mappings);
+}
+
+static void _register_atexit_callback(void)
+{
+    static oe_once_t once = OE_ONCE_INITIALIZER;
+    oe_once(&once, _call_atexit);
+}
+
+static oe_result_t _validate_mmap_parameters(
+    void* addr,
+    size_t length,
+    int prot,
+    int flags,
+    int fd,
+    off_t offset)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    int flags_copy = flags;
+
+    oe_errno = OE_EINVAL;
+
+    // If addr is not NULL, then the kernel takes it as a hint about where to
+    // place the mapping; on     Linux, the kernel will pick a nearby page
+    // boundary (but always above or equal to the value specified by
+    //   /proc/sys/vm/mmap_min_addr) and attempt to create the mapping there.
+    // OE currently does not support this usage.
+    if (addr != NULL)
+        OE_RAISE_MSG(
+            OE_UNSUPPORTED, "non-zero `addr` parameter is not supported.");
+
+    // PROT_NONE and PROT_EXEC are not supported.
+    if (prot == PROT_NONE || (prot & PROT_EXEC))
```

Comment:
How do we enforce `PROT_READ` or `PROT_WRITE` only?

Reply:
Currently we don't enforce this by design.
In SGX1, we'd need to make an OCALL to call mprotect on the page to change the permissions.
However, the enclave has no way to enforce that the host has honored the mprotect request.
There this does not seem possible for SGX1.

There is also the added complexity that changing page permissions could corrupt the allocator since the allocator may choose to store house-keeping data on freed pages. To reliably support coexistence with the allocator, we would need to move mmap to live beneath the allocator.